core/shared/src/main/scala/scalaz/zio/IO.scala

```diff
@@ -161,6 +162,18 @@ sealed abstract class IO[+E, +A] extends Serializable { self =>
         )
     )
 
+  /**
+   * Races this action with the specified action, returning the first
+   * result to *finish*, whether with by producing a value or by failing
+   * with an error. If either of two actions fails before the other succeeds,
+   * the whole race will fail with that error.
+   */
+  final def raceAttempt[E1 >: E, B](that: IO[E1, B]): IO[E1, Either[A, B]] =
+    raceWith(that)(
+      { case (l, f) => l.fold(f.interrupt *> IO.fail0(_), a => IO.now(Left(a))) },
+      { case (r, f) => r.fold(f.interrupt *> IO.fail0(_), a => IO.now(Right(a))) }
```

Comment:
For slightly better performance, consider `IO.nowLeft` and `IO.nowRight`, which avoid the overhead of creating a closure.